### ISSUE TITLE
VirtualRouter not matching prefix

### ISSUE THREAD

**Issue by benpettman** [aws/aws-app-mesh-roadmap#223]
I have been working with all of the demos and walkthroughs, and have applied all of the correct logic to my app.

Upon deploying the app I have a VirtualRouter below

```
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualRouter
metadata:
  namespace: search
  name: api
spec:
  listeners:
    - portMapping:
        port: 7000
        protocol: http
  routes:
    - name: api-route
      httpRoute:
        match:
          prefix: /search
        action:
          weightedTargets:
            - virtualNodeRef:
                name: search-v1
              weight: 1
            - virtualNodeRef:
                name: search-v2
              weight: 1
```

As you can see I am wanting to only apply this weighted routing on the path ```/search``` however when I curl into a curler pod into the EKS cluster and curl the ```/health``` endpoint this gets balanced as per the weightings above.

I would of expected it to only weight the requests going to ```/search``` and not ```/health```, or am I doing something wrong or mis-understanding?

EDIT
My controller is spitting out the following log

```
2020-07-02T10:19:57.92926432Z {"level":"error","ts":1593685197.9289563,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"virtualservice","request":"search/api.search.travelverse.local","error":"failed to resolve virtualRouterRef: unable to fetch virtualRouter: search/api-router: VirtualRouter.appmesh.k8s.aws \"api-router\" not found","errorVerbose":"VirtualRouter.appmesh.k8s.aws \"api-router\" not found\nunable to fetch virtualRouter: search/api-router\ngithub.com/aws/aws-app-mesh-controller-for-k8s/pkg/references.(*defaultResolver).ResolveVirtualRouterReference\n\t/workspace/pkg/references/resolver.go:97\ngithub.com/aws/aws-app-mesh-controller-for-k8s/pkg/virtualservice.(*defaultResourceManager).findVirtualRouterDependencies\n\t/workspace/pkg/virtualservice/resource_manager.go:172\ngithub.com/aws/aws-app-mesh-controller-for-k8s/pkg/virtualservice.(*defaultResourceManager).Reconcile\n\t/workspace/pkg/virtualservice/resource_manager.go:74\ngithub.com/aws/aws-app-mesh-controller-for-k8s/controllers/appmesh.(*virtualServiceReconciler).reconcileVirtualService\n\t/workspace/controllers/appmesh/virtualservice_controller.go:102\ngithub.com/aws/aws-app-mesh-controller-for-k8s/controllers/appmesh.(*virtualServiceReconciler).reconcile\n\t/workspace/controllers/appmesh/virtualservice_controller.go:95\ngithub.com/aws/aws-app-mesh-controller-for-k8s/controllers/appmesh.(*virtualServiceReconciler).Reconcile\n\t/workspace/controllers/appmesh/virtualservice_controller.go:67\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:256\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:90\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1373\nfailed to resolve virtualRouterRef\ngithub.com/aws/aws-app-mesh-controller-for-k8s/pkg/virtualservice.(*defaultResourceManager).findVirtualRouterDependencies\n\t/workspace/pkg/virtualservice/resource_manager.go:174\ngithub.com/aws/aws-app-mesh-controller-for-k8s/pkg/virtualservice.(*defaultResourceManager).Reconcile\n\t/workspace/pkg/virtualservice/resource_manager.go:74\ngithub.com/aws/aws-app-mesh-controller-for-k8s/controllers/appmesh.(*virtualServiceReconciler).reconcileVirtualService\n\t/workspace/controllers/appmesh/virtualservice_controller.go:102\ngithub.com/aws/aws-app-mesh-controller-for-k8s/controllers/appmesh.(*virtualServiceReconciler).reconcile\n\t/workspace/controllers/appmesh/virtualservice_controller.go:95\ngithub.com/aws/aws-app-mesh-controller-for-k8s/controllers/appmesh.(*virtualServiceReconciler).Reconcile\n\t/workspace/controllers/appmesh/virtualservice_controller.go:67\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:256\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:90\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1373","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:258\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.18.2/pkg/util/wait/wait.go:90"}
```
Which would seem to say it cant find the virtual router, if this is the case then how would it be using the router to effectively load balance me over the 2 virtual nodes?

EDIT 2
What is odd about that error is it cant find ```api-router``` by my route isnt called that, its called ```api```

**Comment by benpettman** [comment]
Ive since then amended my virtual router to the below

```
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualRouter
metadata:
  namespace: search
  name: api
spec:
  listeners:
    - portMapping:
        port: 7000
        protocol: http
  routes:
    - name: health-route
      priority: 1
      httpRoute:
        match:
          prefix: /health
        action:
          weightedTargets:
            - virtualNodeRef:
                name: search-v1
              weight: 1
    - name: api-route
      priority: 10
      httpRoute:
        match:
          prefix: /search
        action:
          weightedTargets:
            - virtualNodeRef:
                name: search-v1
              weight: 1
            - virtualNodeRef:
                name: search-v2
              weight: 1
```

And still my routes are being ignored, when I go to ```/health``` it is still doing a round robin on the ```api-route```

**Comment by benpettman** [comment]
Removed the whole route named ```api-route``` and its still doing a round robin some how? Is there a cache that needs to be cleared?

**Comment by benpettman** [comment]
I have replicated this with the colorapp demo, removed the green one from the router and then did a curl to the ```color.search.svc.cluster.local``` and it still returns green and blue. 

I dont understand why its not being proxied at all?

**Comment by rajal-amzn** [comment]
@benpettman When you update a route, it takes couple of minutes for the Envoy to receive the updated configuration.

But, for the example you mentioned, "/health" should not redirect the request to "/search". Can you share your full mesh configuration for us to reproduce the issue?

**Comment by benpettman** [comment]
Sure, I mean Ill share the demo colorapp as I can replicate the same issues with that:

```
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualNode
metadata:
  name: blue
  namespace: search
spec:
  podSelector:
    matchLabels:
      app: color
      version: blue
  listeners:
    - portMapping:
        port: 8080
        protocol: http
      healthCheck:
        protocol: http
        path: '/ping'
        healthyThreshold: 2
        unhealthyThreshold: 2
        timeoutMillis: 2000
        intervalMillis: 5000
  serviceDiscovery:
    dns:
      hostname: color-blue.search.svc.cluster.local
---
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualNode
metadata:
  name: green
  namespace: search
spec:
  podSelector:
    matchLabels:
      app: color
      version: green
  listeners:
    - portMapping:
        port: 8080
        protocol: http
      healthCheck:
        protocol: http
        path: '/ping'
        healthyThreshold: 2
        unhealthyThreshold: 2
        timeoutMillis: 2000
        intervalMillis: 5000
  serviceDiscovery:
    dns:
      hostname: color-green.search.svc.cluster.local
---
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualService
metadata:
  name: color
  namespace: search
spec:
  awsName: color.search.svc.cluster.local
  provider:
    virtualRouter:
      virtualRouterRef:
        name: color
---
apiVersion: appmesh.k8s.aws/v1beta2
kind: VirtualRouter
metadata:
  namespace: search
  name: color
spec:
  listeners:
    - portMapping:
        port: 8080
        protocol: http
  routes:
    - name: color-route
      httpRoute:
        match:
          prefix: /
        action:
          weightedTargets:
            - virtualNodeRef:
                name: blue
              weight: 1
---
apiVersion: v1
kind: Service
metadata:
  name: color-green
  namespace: search
spec:
  ports:
    - port: 8080
      name: http
  selector:
    app: color
    version: green
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: green
  namespace: search
  labels:
    app: color
    version: green
spec:
  replicas: 1
  selector:
    matchLabels:
      app: color
      version: green
  template:
    metadata:
      labels:
        app: color
        version: green
    spec:
      containers:
        - name: app
          image: 327913457104.dkr.ecr.eu-west-1.amazonaws.com/appmesh-demo/colorapp
          ports:
            - name: app-port
              containerPort: 8080
          readinessProbe:
            httpGet:
              path: /ping
              port: 8080
            initialDelaySeconds: 1
            periodSeconds: 5
            failureThreshold: 1
          livenessProbe:
            httpGet:
              path: /ping
              port: 8080
            initialDelaySeconds: 1
            periodSeconds: 5
            failureThreshold: 1
          env:
            - name: "COLOR"
              value: "green"
---
apiVersion: v1
kind: Service
metadata:
  name: color-blue
  namespace: search
spec:
  ports:
    - port: 8080
      name: http
  selector:
    app: color
    version: blue
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: blue
  namespace: search
spec:
  replicas: 1
  selector:
    matchLabels:
      app: color
      version: blue
  template:
    metadata:
      labels:
        app: color
        version: blue
    spec:
      containers:
        - name: app
          image: 327913457104.dkr.ecr.eu-west-1.amazonaws.com/appmesh-demo/colorapp
          ports:
            - name: app-port
              containerPort: 8080
          readinessProbe:
            httpGet:
              path: /ping
              port: 8080
            initialDelaySeconds: 1
            periodSeconds: 5
            failureThreshold: 1
          livenessProbe:
            httpGet:
              path: /ping
              port: 8080
            initialDelaySeconds: 1
            periodSeconds: 5
            failureThreshold: 1
          env:
            - name: "COLOR"
              value: "blue"
---
apiVersion: v1
kind: Service
metadata:
  name: color
  namespace: search
  labels:
    app: color
spec:
  ports:
    - port: 8080
      name: http
  selector:
    app: color
```

Feel free to tell me its wrong!

After deploying that, and then deploying a curler pod and calling ``` curl color.search.svc.cluster.local ``` I get blue and green, where I think it should just get blue as the VR only has blue in it.

**Comment by benpettman** [comment]
Here is a curl result from the above config...

```
root@curler:/# curl color.search.svc.cluster.local:8080
green
root@curler:/# curl color.search.svc.cluster.local:8080
green
root@curler:/# curl color.search.svc.cluster.local:8080
blue
root@curler:/# curl color.search.svc.cluster.local:8080
blue
root@curler:/# curl color.search.svc.cluster.local:8080
blue
root@curler:/# curl color.search.svc.cluster.local:8080
blue
root@curler:/# curl color.search.svc.cluster.local:8080
green
root@curler:/# curl color.search.svc.cluster.local:8080
green
root@curler:/# curl color.search.svc.cluster.local:8080
blue
root@curler:/# curl color.search.svc.cluster.local:8080
green
root@curler:/# curl color.search.svc.cluster.local:8080
blue
root@curler:/#
```

**Comment by benpettman** [comment]
I think I have worked it out, the mesh doesnt work if you dont call something from within the mesh.

facepalm

**Comment by rajal-amzn** [comment]
That makes sense. Currently, services can only call other services within the mesh. We are tracking support for ingress [here](https://github.com/aws/aws-app-mesh-roadmap/issues/37)
Also, I note that you are using the older version of K8s controller. If you havent checked already, we have released v1.0 of K8s [here](https://github.com/aws/aws-app-mesh-controller-for-k8s/releases/tag/v1.0.0) which fixes lot of bugs and provides improved user experience.
